File: 535.逆序对/Solution.py
class Solution:
    # @param {int[]} A an array
    # @return {int} total of reverse pairs
    def reversePairs(self, A):
        # Write your code here
        A, rlt = self.merge_sort(A)
        return rlt
        
    def merge_sort(self, A):
        n = len(A)
        count = 0
        if n in [0, 1]:
            return A, count
        left, countl = self.merge_sort(A[0:n//2])
        count += countl
        right, countr= self.merge_sort(A[n//2:])
        count += countr
        A = []
        i, j = 0, 0
        while i < len(left) and j < len(right):
            if left[i] <= right[j]:
                A.append(left[i])
                i += 1
            else:
                A.append(right[j])
                j += 1
                count += len(left) - i
        A = A + left[i:] + right[j:]
        return A, count

File: 535.逆序对/Solution.cpp
class Solution {
public:
    /**
     * @param A an array
     * @return total of reverse pairs
     */
    long long reversePairs(vector<int>& A) {
        // Write your code here
        long long count = 0;
        merge_sort(A, count);
        return count;
    }
    
    void merge_sort(vector<int>& A, long long& count){
        if(A.size() <= 1) return;
        
        int n = A.size()/2;
        vector<int> left(A.begin(), A.begin()+n);
        vector<int> right(A.begin()+n, A.end());
        merge_sort(left, count);
        merge_sort(right, count);
        
        A.clear();
        while(left.size() > 0 && right.size() > 0){
           if(left[0] > right[0]) {
               count = count + left.size();
               A.push_back(right[0]);
               right.erase(right.begin());
           }
           else{
               A.push_back(left[0]);
               left.erase(left.begin());
           }
        }
        while(left.size() > 0){
            A.push_back(left[0]);
            left.erase(left.begin());
        }
        while(right.size() > 0){
            A.push_back(right[0]);
            right.erase(right.begin());
        }
        return;   
    }
};


















